session.go

```diff
@@ -133,7 +139,10 @@ type session struct {
 
 	closeOnce sync.Once
 	// closeChan is used to notify the run loop that it should terminate
-	closeChan chan closeError
+	closeChan   chan closeError
+	delayedChan chan delayedError
```

Comment:
I think you could avoid the need for this channel by sending insecure close errors over closeChan, appropriately marked, and checking those errors when they're received in the mainloop instead of `break`ing unconditionally.